File: julia/Other/advent_of_code/2021/day1/day1.zig
const std = @import("std");
const fs = std.fs;
const io = std.io;
const ArrayList = std.ArrayList;
// const testing_allocator = std.testing.allocator;
const parseInt = std.fmt.parseInt;

pub fn main() !void {
    // Parse input
    var file = try fs.cwd().openFile("data1.txt", .{});
    defer file.close();
    var buf_reader = io.bufferedReader(file.reader());
    var in_stream = buf_reader.reader();
    var buf: [1024]u8 = undefined;
    var arena_state = std.heap.ArenaAllocator.init(std.heap.c_allocator);
    defer arena_state.deinit();
    const allocator = arena_state.allocator();
    var data = ArrayList(isize).init(allocator);
    defer data.deinit();
    while (try in_stream.readUntilDelimiterOrEof(&buf, '\n')) |line| {
        try data.append((try parseInt(isize, line, 10)));
    }
    // std.debug.print("{}", .{data});

    // Part 1
    const part1_solution = part1(data);
    std.debug.print("Part 1: {}\n", .{part1_solution});

    // Part 2
    const part2_solution = part2(data);
    std.debug.print("Part 2: {}\n", .{part2_solution});
}

fn part1(data: ArrayList(isize)) isize {
    var cnt: isize = 0;
    {
        var i: usize = 1;
        while (i < data.items.len) : (i += 1) {
            if (data.items[i - 1] < data.items[i]) {
                cnt += 1;
            }
        }
    }
    return cnt;
}

fn part2(data: ArrayList(isize)) isize {
    var cnt: isize = 0;
    {
        var i: usize = 3;
        while (i < data.items.len) : (i += 1) {
            const a: isize = data.items[i - 1] + data.items[i - 2] + data.items[i - 3];
            const b: isize = data.items[i] + data.items[i - 1] + data.items[i - 2];
            if (a < b) {
                cnt += 1;
            }
        }
    }
    return cnt;
}

File: julia/Other/advent_of_code/2021/day1/day1.c
#include <stdio.h>

// Number of lines of input (data1.txt)
#define N 2000

// Length of input
int data_sz = 0;

int part1(int data[])
{
	int cnt = 0;
	for (int i = 1; i < data_sz; ++i) {
		if (data[i - 1] < data[i]) {
			++cnt;
		}
	}
	return cnt;
}

int part2(int data[])
{
	int cnt = 0;
	for (int i = 3; i < data_sz; ++i) {
		int a = data[i - 1] + data[i - 2] + data[i - 3];
		int b = data[i] + data[i - 1] + data[i - 2];
		if (a < b) {
			cnt += 1;
		}
	}
	return cnt;
}

int main()
{
	// Open file
	FILE *f;
	f = fopen("data1.txt", "r");

	if (f == 0) {
		printf("Could not open file.\n");
		return 1;
	}

	// Read file into array
	int data[N];
	data_sz = 0;

	while (!feof(f)) {
		int x = 0;
		int n = fscanf(f, "%d\n", &x);
		if (n == 1) {
			data[data_sz++] = x;
		}
	}

	// Close file
	fclose(f);

	// Display results!
	int part1_solution = part1(data);
	printf("Part 1: %d\n", part1_solution);
	int part2_solution = part2(data);
	printf("Part 2: %d\n", part2_solution);
}
